### ISSUE TITLE
Behavior change in 10.0.0 - getStreamData(level=qpdf_dl_generalized) decompresses JPEGs

### ISSUE THREAD

**Issue by jbarlow83** [qpdf/qpdf#425]
As the title suggests, `QPDFObjectHandle::getStreamData()` will now decompress JPEGs, even at the default value of `qpdf_dl_generalized`. (Not sure if it matters, but I call `QPDF::immediateCopyFrom(true)` on all PDFs.)

If I'm not mistaken this change was not intended, or if it was the documentation does not seem to reflect this.

pikepdf's test suite assumes JPEGs will not be decoded this way.

**Comment by jberkenbilt** [comment]
You are correct. It was not intended. I will fix for 10.0.1.

**Comment by jberkenbilt** [comment]
@jbarlow83 I'm not able to reproduce this. Can you possibly provide a stand-alone case?

**Comment by jberkenbilt** [comment]
Assuming it is broken, 893d38b87e4ad6c6c55f49464f6b721c516ec878 is the most likely culprit. While I was working on that, I ran into a few instances of things either being compressed twice or being compressed not at all, but I added extra test cases and found the logic error. Or at least I thought I did.

**Comment by jberkenbilt** [comment]
For what it's worth, here was my attempt at reproducing the problem. This code hard-codes /tmp/a.pdf. My /tmp/a.pdf is attached.

```c++
#include <qpdf/QPDF.hh>
#include <qpdf/QPDFObjectHandle.hh>
#include <qpdf/QUtil.hh>

int main()
{
    QPDF qpdf;
    qpdf.setImmediateCopyFrom(true);
    qpdf.processFile("/tmp/a.pdf");
    QPDFObjectHandle o = qpdf.getObjectByID(6, 0);
    auto b = o.getStreamData();
    std::cout << b->getSize() << std::endl;
    auto buf = b->getBuffer();
    std::string data(reinterpret_cast<char*>(buf), 10);
    std::cout << QUtil::hex_encode(data) << std::endl;
    QPDF other;
    other.setImmediateCopyFrom(true);
    other.emptyPDF();
    auto ocopy = other.copyForeignObject(o);
    b = ocopy.getStreamData();
    std::cout << b->getSize() << std::endl;
    buf = b->getBuffer();
    data = std::string(reinterpret_cast<char*>(buf), 10);
    std::cout << QUtil::hex_encode(data) << std::endl;
    return 0;
}
```
[a.pdf](https://github.com/qpdf/qpdf/files/4450689/a.pdf)

**Comment by jbarlow83** [comment]
I reproduced it with your example.

Using your code, linked against qpdf 9.1.1:

```shell
$ g++ -g -O2 -std=c++11 issue425.cpp -o issue425_system_lqpdf -lqpdf
$ ./issue425_system_lqpdf
libc++abi.dylib: terminating with uncaught exception of type QPDFExc: /tmp/a.pdf (offset 623): getStreamData called on unfilterable stream
fish: './issue425_system_lqpdf' terminated by signal SIGABRT (Abort)
```

Since object 6,0 is a JPEG, this exception and abort is the expected behavior.

Linked against qpdf 10 (release-qpdf-10.0.0-1-g496aacd):

```shell
$ g++ -g -O2 -std=c++11 issue425.cpp -o issue425_new  -I (pwd)/include libqpdf/build/.libs/libqpdf.a -ljpeg -lz -lgnutls
$ ./issue425_new
1770
ffd8ffe000104a464946
1770
ffd8ffe000104a464946
```

This decompressed the JPEG - not expected behavior.

This was on my mac and `g++` is really clang, if that matters...

In pikepdf the reproducing test is

```
pytest tests/test_image_access.py::test_extract_direct_fails_nondefault_colortransform
```

which amounts to

```python
import pikepdf
p = pikepdf.open('a.pdf')
try:
    p.pages[0].images['/Im1'].read_bytes()
except pikepdf.PdfError:
    print("success")
else:
    print("failed")
```

**Comment by jberkenbilt** [comment]
That hex string is not the uncompressed jpeg. It is the compressed JPEG. But I know what the problem is. I goofed up in switching of the pipeStreamData API, which is backward compatible, but I switched internally to the new API, so it's falling back to returning uncompressed data rather than failing. At least that's what I think. Should be a one-line fix. I'll try to get 10.0.1 out tomorrow.

Anyway, the bug isn't that it's uncompressing JPEGs, the bug is that it's returning the raw data instead of failing. Thanks the report. This is a pretty serious error, so I'm glad to be able to get it fixed in the first few days.
